### PR TITLE
Initial bors-ng integration

### LINKED ISSUE #324: Investigate Bors
Discussion over on https://gcc-rust.zulipchat.com/#narrow/stream/266897-general/topic/GitHub.20repository.20write.20access.3F/near/232551949

Raised the idea of using bors:

- https://bors.tech/
- https://graydon.livejournal.com/186550.html

### PR DESCRIPTION
**Pull request by dkm**
Github Action only triggered on bors-ng controlled branches (trying/staging).
Fixes #324

### CODE REVIEW

**Comment by dkm** [comment]
This is only a part of what is needed. The bors-ng app must first be added to the repository and some operations on master branch could be limited

**Comment by dkm** [comment]
as example, see : 

- example of "try" command (CI job executed for a PR) : https://github.com/dkm/gccrs/pull/3
- example of "r+" command (bors applies the PR on master, executes CI and merge if successful) : https://github.com/dkm/gccrs/pull/1

**Comment by philberty** [comment]
I really like the look of this. Will we go for it? @tschwinge as this project is destined for GCC someday I would like to be mindful of that and we have to be careful about no copyright assignments etc so the automating merging is kind of scary in that regard.

Are there limitations here to take into account for this?

**Comment by dkm** [comment]
That's not full automatic, that's simply more automated. You still need someone to ask bors to do the job. It's similar to having someone click the "rebase and merge" button, but you are sure the merge only happens iff CI is good.

**Comment by flip1995** [comment]
Is your concern that the bot has no copyright assignment, but will produce merge commits?

Or that it may merge PRs from people that don't have copyright assignment in place?

You can check the second with a CI job that is run by bors, as long as the FSF provides an api to look up copyright assignment (it would be kind of ridiculous if it doesn't).

File: .github/workflows/ccpp.yml
Comment (flip1995) [inline .github/workflows/ccpp.yml:6]: You still want to run CI on pull requests.

**Comment by dkm** [comment]
The setup tested is really basic. It means that the workflow would be more or less like this:
- new PR opened => no CI job (we could still run some linter as it's lightweight?)
- reviewers comments and at some point decide:
 -  it's worth testing : "bors try" => CI job executed, result posted in the PR, nothing more
 -  it's ok to merge iff CI is fine: "bors r+" => CI job executed, if all is OK, a merge commit is created

I think you can do more complex setup. I can continue the readings and see how other projects are doing (I guess @bjorn3 or @flip1995 have better knowledge on this !)

**Comment by philberty** [comment]
What about also creating a commit message format check option 

> Is your concern that the bot has no copyright assignment, but will produce merge commits?
> 
> Or that it may merge PRs from people that don't have copyright assignment in place?
> 
> You can check the second with a CI job that is run by bors, as long as the FSF provides an api to look up copyright assignment (it would be kind of ridiculous if it doesn't).

Thanks, @flip1995 yeah my concern was if it was merging automatically that it would merge stuff from people with no copyright assignment. I will ask David Edelsohn on Monday if there is an API to lookup copyright assignments but I doubt it. That would be a really cool automation to have.

File: .github/workflows/ccpp.yml
Comment (dkm) [reply .github/workflows/ccpp.yml:6]: Wasn't really sure what we wanted. I had this at the beginning but had the impression that doing any bors command would still rerun the job, making this initial job a bit less useful. But I may have overlooked this !

**Comment by flip1995** [comment]
> - new PR opened => no CI job (we could still run some linter as it's lightweight?)

You definitely want to run (most of) your CI on PRs.

**Comment by dkm** [comment]
> What about also creating a commit message format check option

That would prevent WIP commit being merged at least (sorry for this).
What would you enforce then? Do we want to have GNU changelog & commit message already?

**Comment by flip1995** [comment]
>  my concern was if it was merging automatically that it would merge stuff from people with no copyright assignment. 

A reviewer will still have to accept the PR. And you select who the reviewers are. So this shouldn't be a problem. 

> I will ask David Edelsohn on Monday if there is an API to lookup copyright assignments but I doubt it. 

I would expect that something like this exists. Putting such a big contributon restriction in place and then not being able to check it automatically sounds stupid to me TBH.

**Comment by flip1995** [comment]
> To check whether papers have been received, look in `/gd/gnuorg/copyright.list`. If you can’t look there directly, `fsf-records@gnu.org` can check for you. Our clerk can also check for papers that are waiting to be entered and inform you when expected papers arrive.

> The directory `/gd/gnuorg` mentioned throughout this document is available on the general GNU server, currently `fencepost.gnu.org`. If you are the maintainer of a GNU package, you should have an account there. If you don’t have one already, see https://www.gnu.org/software/README.accounts.html. Such GNU login accounts include email (see https://www.fsf.org/about/systems/sending-mail-via-fencepost).

So if we'd be able to get an account there for the project, we should be able to set up a CI job that checks this automatically with the login data stored in GHA secrets.

**Comment by tschwinge** [comment]
Regarding FSF copyright assignment status:

Indeed I'm not aware of an API for retrieving that.  On the very few occasions where I have to check this, I look up `~gnuorg/copyright.list` on `fencepost.gnu.org` (I do have an account).

GCC upstream solves this problem by having people with GCC repository access privileges and FSF copyright assignment status verified list themselves in the `MAINTAINERS` file, *Write After Approval* section.  (You gain that status, basically, by submitting and getting approved one sensible patch.)  For the "random one-off contributor" (without GCC repository access privileges), FSF copyright assignment status is manually verified (if applicable by the nature of the patch submitted).  So, this basically works on a trust basis.

Now, if we really want to implement strict checking, I'd suggest we add an additional file next to `MAINTAINERS`, (manually) in there listing the GitHub handles where we have (manually) verified the FSF copyright assignment status, and have the GitHub CI bot check that.

**Comment by tschwinge** [comment]
> > What about also creating a commit message format check option
> 
> That would prevent WIP commit being merged at least (sorry for this).
> What would you enforce then? Do we want to have GNU changelog & commit message already?

I don't think that's useful at this point: eventually, we're going to squash all the commits anyway (so the `ChangeLog` files then need to change accordingly), and for our own ongoing development, the GNU `ChangeLog` format doesn't add valuable information (in my opinion).  Instead it should be "obvious" that everyone come up with some kind of reasonable Git commit message (as we're indeed already doing).

File: .github/workflows/ccpp.yml
Comment (flip1995) [reply .github/workflows/ccpp.yml:6]: The bors command is usually only triggered exactly once and that is right before the PR is merged with `bors r+`. bors doesn't replace your CI. It's only there to automate checking PRs before merging.

Comment (flip1995) [reply .github/workflows/ccpp.yml:6]: You almost never use `bors try`. The only scenario where you use `bors try` is when you're `bors` CI runs more checks than your PR CI and you want to check if a WIP PR would already pass the `bors` CI.

Comment (dkm) [reply .github/workflows/ccpp.yml:6]: Ok, so I changed it to follow your advice. Everything is executed when a PR is opened and we have no extra tests for bors. Thanks for the comments!

**Comment by dkm** [comment]
Updated details:

> The setup tested is really basic. It means that the workflow would be more or less like this:
> 
>     * new PR opened => CI job executed
> 
>     * reviewers comments and at some point decide:
> 
>     * it's ok to merge iff CI is fine: "bors r+" => bors apply changes to latest master, CI job executed, if all is OK, a merge commit is created

What should be configured at the same time: (see https://bors.tech/documentation/getting-started/ )
- install bors-ng app for the gccrs repository (takes 5min, only need few clicks): https://github.com/apps/bors 
- maybe restrict the manual merge to avoid errors. I don't know how bigger projects are doing. Maybe only some dev have the permission to merge PR directly in master ?

**Comment by philberty** [comment]
Will we give this a go and merge it this week?

**Review by philberty — APPROVED** [review]

**Review by flip1995 — APPROVED** [review]
LGTM

**Comment by dkm** [comment]
> Will we give this a go and merge it this week?

I don't think I can prepare anything more, only you can add the bors-ng app and maybe restrict what "regular" reviewers can do directly in master branch :)

**Review by NalaGinrut — APPROVED** [review]
LGTM

**Review by SimplyTheOther — APPROVED** [review]

**Pull request merged**
